File: src/037/p037.py
import itertools
from src.utils.py.prime import generate_primes, is_prime


def is_left_truncatable(digits: tuple[int, ...], primes: list[int]) -> bool:
    string = "".join(str(d) for d in digits)
    for i in range(len(digits)):
        if not is_prime(int(string[i:]), primes=primes):
            return False
    return True


def is_right_truncatable(digits: tuple[int, ...], primes: list[int]) -> bool:
    string = "".join(str(d) for d in digits)
    for i in range(len(digits)):
        if not is_prime(int(string[: i + 1]), primes=primes):
            return False
    return True


def solve() -> int:
    primes = generate_primes(1000)
    truncatable_primes: list[int] = []
    for length in itertools.count():
        if len(truncatable_primes) == 11:
            break

        middle_digits = itertools.tee([1, 3, 7, 9], length)
        for combo in itertools.product([1, 2, 3, 5, 7, 9], *middle_digits, [3, 7]):
            if is_left_truncatable(combo, primes=primes) and is_right_truncatable(
                combo, primes=primes
            ):
                truncatable_primes.append(int("".join(str(d) for d in combo)))
    return sum(truncatable_primes)

File: src/037/p037.go
package proble037

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/sliceutils"
	"github.com/sunilbpandey/project-euler/src/utils/go/strutils"
)

func isPrime(s string, primes []int) bool {
	return primeutils.IsPrime(strutils.Atoi(s), primes)
}

func isTruncatablePrime(n int, primes []int) bool {
	s := strconv.Itoa(n)
	for i := 0; i < len(s); i++ {
		if !isPrime(s[:len(s)-i], primes) || !isPrime(s[i:], primes) {
			return false
		}
	}
	return true
}

func Solve() string {
	primes := primeutils.GeneratePrimes(1000)
	sum := 0
	for length, count := 0, 0; count < 11; length++ {
		for _, prefix := range []int{2, 3, 5, 7} {
			for middle := range sliceutils.Product([]int{1, 3, 7, 9}, length) {
				for _, suffix := range []int{3, 7} {
					digits := append(append([]int{prefix}, middle...), suffix)
					n := strutils.Atoi(strutils.JoinInts(digits))
					if isTruncatablePrime(n, primes) {
						sum += n
						count += 1
					}
				}
			}
		}
	}
	return strconv.Itoa(sum)
}
